### ISSUE TITLE
[Feature Request] Implement Range Requests for <video> / <audio>

### ISSUE THREAD

**Issue by bogeychan** [elysiajs/elysia-static#19]
Due to the current implementation, the frontend must download the entire file, which in turn prevents _buffering_ ("the frontend has to download the whole file first") in the [video](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/video) / [audio](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/audio) control:

https://github.com/elysiajs/elysia-static/blob/403223119bdc52910c75044b7cf5ebf4bda01699/src/index.ts#L160

Bun supports [Streaming files](https://bun.sh/docs/api/http#streaming-files):

> You can send part of a file using the `slice(start, end)` method on the `Bun.file` object. This automatically sets the `Content-Range` and `Content-Length` headers on the Response object.

However this behavior doesn't allow _seeking_ ("jumping to a timestamp") because the _actual size_ is not set in the `Content-Range` header:

- Bun: `bytes 0-10/*`
- Required: `bytes 0-10/20`

> manually setting the header works but it ends up twice in the response (with `bun@1.0.11` - due to `slice(start, end)`)
> ```ts
> new Response(Bun.file(filename).slice(start, end), {
>   status: 206,
>   headers: {
>     'Content-Range': `bytes ${start}-${end - 1}/${size}`
>   }
> });
> ```
> [HTTP range requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests)

**Comment by libersoft-org** [comment]
I've just found a walkaround:

```
async getDownload(req) {
if (!req.params.hash || !req.params.name) return this.getIndex(req);
const file = Bun.file(path.join(Common.settings.storage.download, req.params.hash));
if (!await file.exists()) return this.getIndex(req);
if (req.headers.range) {
 const chunk = Common.settings.storage.chunk_download;
 let [start = 0, end = Infinity] = req.headers.range.split('=').at(-1).split('-').map(Number);
 if (end == 0) end = start + chunk < file.size ? start + chunk : file.size;
 return new Response(file.slice(start, start + chunk), {
  status: 206,
  headers: { 'Content-Range': 'bytes ' + start + '-' + end + '/' + file.size }
 });
} else {
 return new Response(file, {
  headers: {
   'Content-Type': 'application/octet-stream',
   'Content-Disposition': 'attachment; filename="' + req.params.name + '"'
  }
 });
}
```

**Comment by libersoft-org** [comment]
I've just found a walkaround:

```javascript
 async getDownload(req) {
  if (!req.params.hash || !req.params.name) return this.getIndex(req);
  const file = Bun.file(path.join(Common.settings.storage.download, req.params.hash));
  if (!await file.exists()) return this.getIndex(req);
  if (req.headers.range) {
   const chunk = Common.settings.storage.chunk_download;
   let [start = 0, end = Infinity] = req.headers.range.split('=').at(-1).split('-').map(Number);
   if (end == 0) end = start + chunk < file.size ? start + chunk : file.size;
   return new Response(file.slice(start, start + chunk), {
    status: 206,
    headers: { 'Content-Range': 'bytes ' + start + '-' + end + '/' + file.size }
   });
  } else {
   return new Response(file, {
    headers: {
     'Content-Type': 'application/octet-stream',
     'Content-Disposition': 'attachment; filename="' + req.params.name + '"'
    }
   });
  }

**Comment by libersoft-org** [comment]
> I've just found a walkaround:
> 
> ```js
>  async getDownload(req) {
>   if (!req.params.hash || !req.params.name) return this.getIndex(req);
>   const file = Bun.file(path.join(Common.settings.storage.download, req.params.hash));
>   if (!await file.exists()) return this.getIndex(req);
>   if (req.headers.range) {
>    const chunk = Common.settings.storage.chunk_download;
>    let [start = 0, end = Infinity] = req.headers.range.split('=').at(-1).split('-').map(Number);
>    if (end == 0) end = start + chunk < file.size ? start + chunk : file.size;
>    return new Response(file.slice(start, start + chunk), {
>     status: 206,
>     headers: { 'Content-Range': 'bytes ' + start + '-' + end + '/' + file.size }
>    });
>   } else {
>    return new Response(file, {
>     headers: {
>      'Content-Type': 'application/octet-stream',
>      'Content-Disposition': 'attachment; filename="' + req.params.name + '"'
>     }
>    });
>   }
> ```

Hmm, this walkaround seems to be working in Firefox only, not in Chrome. The problem is that the header range is present twice :(

**Comment by bogeychan** [comment]
blocked by https://github.com/oven-sh/bun/issues/7051

**Comment by hckhanh** [comment]
https://github.com/oven-sh/bun/issues/7051 is closed

**Comment by SaltyAom** [comment]
Should have been fixed by 4ba3e85 published under 0.8.0.

Feel free to reopen the issue if the problem still persists.
